Automattic/wp-calypso
# Terms: Moving category-selector to term-selector

For #3697- This branch begins the task of transitioning the `<CategorySelector />` to a generic `<TermSelector />` for use with CPT taxonomies.  This has branch has been a bit of a winding trail through the redux forest, but it has now arrived at a happy meadow, and is ready for a review or two 👍 

The final path of this branch involved using much of the same logic that @aduth used in #5647.  The `<CategorySelector />` and `<PostSelector />` have shared many similarities in the past, and this continues here with the adoption of ( much of the same code ) and `ReactVirtualized`.  I wonder if there might be any way to possibly merge these two components since they are quite similar.  Though there are differences, and maybe some magic is best left untouched.

Akin to the `<CategorySelector />`, the first implementation of the `<TermSelector />` is in the menu builder - replacing the current usage of `<CategorySelector />` there.  This is the first bit of redux in this area, so that is kind of fun too!

__To Test__
Essentially to test this branch out, it would be most helpful to have a test site with over 100 categories to play with the InfiniteScroll provided by ReactVirtualized.  Additionally, if you have some nested categories, that is desirable as well.

With copious amounts of test categories available, simply work with the Menu Builder Category section in all crazy ways possible:

![menus_ _testingtimmy2_wordpress_com_ _wordpress_com](https://cloud.githubusercontent.com/assets/22080/16098666/22d7b6bc-3309-11e6-8705-6ce2b24e88b9.png)

Specifically test out the search capability, and infinite scroll.  But basic actions such as selecting a different category, saving, and ensuring the menu updates properly would be excellent testing steps.

__Questions__
There was some chatter around possibly naming this `<HierarchicalTermSelector />` and having a `<TermSelector />` that was smart enough to determine if this component, or the token field should be rendered.  Naming stuff is hard - so your feedback there is appreciated too.  /cc @nylen
